frontends/api/src/hooks/learningResources/index.ts

```diff
@@ -240,6 +244,57 @@ const useOfferorsList = (
   })
 }
 
+/* The `LIST_TYPE_` constants are used to differentiate between
+  different types of lists in the application. In this specific code
+  snippet, `LIST_TYPE_LEARNING_PATH` and `LIST_TYPE_USER_LIST` are
+  used to specify the type of list being operated on when moving
+  items within a list. This helps in determining whether the item
+  should be moved within a learning path or a user list. */
+interface ListItemMoveRequest {
+  listType: string
+  parent: number
+  id: number
+  position?: number
+}
+const useListItemMove = (
+  learningPathListType: string,
+  userListListType: string,
+) => {
+  const queryClient = useQueryClient()
+  return useMutation({
+    mutationFn: ({ listType, parent, id, position }: ListItemMoveRequest) =>
+      new Promise<void>((resolve) => {
+        if (listType === learningPathListType) {
+          learningpathsApi.learningpathsItemsPartialUpdate({
+            learning_resource_id: parent,
+            id,
+            PatchedLearningPathRelationshipRequest: { position },
+          })
+        } else if (listType === userListListType) {
+          userListsApi.userlistsItemsPartialUpdate({
+            userlist_id: parent,
+            id,
+            PatchedUserListRelationshipRequest: { position },
+          })
+        }
+        resolve()
+      }),
```

Comment:
If you try reordering items, you may notice that the UI doesn't update correctly. (Probably not 100% of the time—it depends on the order of some API call returns).

https://github.com/mitodl/mit-open/assets/9010790/4cf6e257-5e77-4b20-a97a-6aeb52d9d9dc

The reason for this issue is fragile code + the `resolve` call above. Note that `resolve` is happening too soon: `learningpathsItemsPartialUpdate` and `userlistsItemsPartialUpdate` are both promises, but the `resolve` call happens before they resolve.

Instead, use `async/await`:
```ts
{
mutationFn: async ({
      listType,
      parent,
      id,
      position,
    }: ListItemMoveRequest) => {
      if (listType === learningPathListType) {
        await learningpathsApi.learningpathsItemsPartialUpdate({
          learning_resource_id: parent,
          id,
          PatchedLearningPathRelationshipRequest: { position },
        })
      } else if (listType === userListListType) {
        await userListsApi.userlistsItemsPartialUpdate({
          userlist_id: parent,
          id,
          PatchedUserListRelationshipRequest: { position },
        })
      }
    },
}
```

------------------------
Here's a fuller explanation of why exactly this was happening:

1. `ItemsListingSortable` maintains a local copy of the sort order: The local order is initially the same as the API order. While you re-order items, the local order changes. When you release the drag, an API call is made and the the database order is updated. If you cancel the drag (press escape) the API call never gets made.

    When new database results come in, the local order is updated:
```tsx
  // ItemListing line 67ish
  /**
   * `sorted` is a local copy of `items`:
   *  - `onSortEnd`, we'll update `sorted` copy immediately to prevent UI from
   *  snapping back to its original position.
   *  - `items` is the source of truth (most likely, this is coming from an API)
   *    so sync `items` -> `sorted` when `items` changes.
   */
  useEffect(() => setSorted(items), [items])
```
2. When you release a drag (submitting a "reorder") `useListItemMove` makes its API call and invalidates the existing query cache...
3. ...so new data gets fetched, updating the local state.

Here the "new data" was getting fetched too early, because the cache was invalidated before the POST call finished.

Reply:
This all makes sense, and to be honest this is the result of me pulling my hair out trying to fix an unrelated linting issue. In `ItemsListingSortable`, where I call `useListItemMove`, it was complaining that `listType` wasn't part of the dependencies. I incorrectly thought that this was an issue with the hook itself, but in reality it was telling me exactly what the problem was. I just forgot to change it back to the `async` / `await` syntax.